File: examples/Actor.Next/04-Auction/Auction.Next.Example04.Tests/AuctionTests.cs
// ------------------------------------------------------------------------
// Copyright 2026 The Dapr Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//     http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ------------------------------------------------------------------------

using Dapr.Actors.Next.Abstractions;
using Dapr.Actors.Next.Abstractions.Options;
using Dapr.Actors.Next.Testing;

namespace Dapr.Actors.Next.Examples.Auction.Tests;

public sealed class AuctionTests
{
    [Fact]
    public async Task Last_second_bid_wins_race_with_close_timer()
    {
        await using var runtime = CreateRuntime(new PriorityActorScheduler(7));
        var auction = runtime.CreateActor<IAuctionActor>(ActorId.Create("auction-1"), "Auction");
        await PlaceBid(runtime, auction, new Bid(100, "alice"));

        runtime.Time.Advance(TimeSpan.FromSeconds(29));
        var lateBid = auction.PlaceBid(new Bid(110, "bob"));
        await runtime.RunToIdle();
        Assert.Equal(BidResult.Accepted, await lateBid);

        runtime.Time.Advance(TimeSpan.FromSeconds(2));
        await runtime.RunToIdle();

        Assert.Equal(AuctionState.Open, await ReadState(runtime, auction));
        Assert.Equal("bob", (await ReadData(runtime, auction)).HighBidder);
    }

    [Fact]
    public void State_machine_has_no_structural_defects()
    {
        ActorStateMachine.Analyze<AuctionActor>().AssertNoStructuralDefects();
        Assert.Contains(ActorStateMachine.Analyze<BadAuctionActor>().StructuralDefects, defect => defect.Contains("unreachable", StringComparison.Ordinal));

        // For deeper exploration, run the same test body under the optional Coyote bridge when enabled.
        Assert.False(CoyoteBridge.IsEnabled);
    }

    private static ActorTestRuntime CreateRuntime(ControlledActorScheduler? scheduler = null)
    {
        _ = typeof(AuctionActor);
        return new ActorTestRuntime(services => services.AddDaprActors(_ => { }), new ActorTestRuntimeOptions { Scheduler = scheduler });
    }

    private static async Task<BidResult> PlaceBid(ActorTestRuntime runtime, IAuctionActor auction, Bid bid)
    {
        var placed = auction.PlaceBid(bid);
        await runtime.RunToIdle();
        return await placed;
    }

    private static async Task<AuctionState> ReadState(ActorTestRuntime runtime, IAuctionActor auction)
    {
        var read = auction.GetState();
        await runtime.RunToIdle();
        return await read;
    }

    private static async Task<AuctionData> ReadData(ActorTestRuntime runtime, IAuctionActor auction)
    {
        var read = auction.GetData();
        await runtime.RunToIdle();
        return await read;
    }
}

File: examples/Actor.Next.FSharp/04-Auction/Auction.Next.FSharp.Example04.Tests/AuctionTests.fs
namespace Auction.Next.FSharp.Example04.Tests

open System
open System.Threading
open System.Threading.Tasks
open Dapr.Actors.Next.Abstractions
open Dapr.Actors.Next.Abstractions.Options
open Dapr.Actors.Next.Testing
open Microsoft.Extensions.DependencyInjection
open Auction.Next.FSharp.Example04
open Xunit

type AuctionTests() =

    static member private CreateRuntime(scheduler: ControlledActorScheduler) =
        // Force-load the Glue assembly and invoke its generated registration module
        // (the module initializer does not fire for LoadFrom in this runtime version)
        let gluePath = System.IO.Path.Combine(AppContext.BaseDirectory, "Auction.Next.FSharp.Example04.Glue.dll")
        let glueAssembly = System.Reflection.Assembly.LoadFrom(gluePath)
        match glueAssembly.GetType("Dapr.Actors.Next.Generated.GeneratedActorRegistrationModule") with
        | null -> failwith "GeneratedActorRegistrationModule not found in Glue assembly"
        | moduleType ->
            match moduleType.GetMethod("Register", System.Reflection.BindingFlags.Static ||| System.Reflection.BindingFlags.NonPublic) with
            | null -> failwith "Register method not found"
            | registerMethod -> registerMethod.Invoke(null, null) |> ignore
        // Force the F# types to be loaded so the generated registry discovers them
        let _ = typeof<AuctionActor>
        let opts = ActorTestRuntimeOptions()
        opts.Scheduler <- scheduler
        ActorTestRuntime(
            (fun services -> services.AddDaprActors(Action<DaprActorsOptions>(fun _ -> ())) |> ignore),
            opts)

    static member private PlaceBid(runtime: ActorTestRuntime, auction: IAuctionActor, bid: Bid) = task {
        let placed = auction.PlaceBid(bid, CancellationToken.None)
        do! runtime.RunToIdle()
        return! placed
    }

    static member private ReadState(runtime: ActorTestRuntime, auction: IAuctionActor) = task {
        let read = auction.GetState(CancellationToken.None)
        do! runtime.RunToIdle()
        return! read
    }

    static member private ReadData(runtime: ActorTestRuntime, auction: IAuctionActor) = task {
        let read = auction.GetData(CancellationToken.None)
        do! runtime.RunToIdle()
        return! read
    }

    [<Fact>]
    member this.Last_second_bid_wins_race_with_close_timer() = task {
        use runtime = AuctionTests.CreateRuntime(PriorityActorScheduler(7))
        let auction = runtime.CreateActor<IAuctionActor>(ActorId.Create("auction-1"), "Auction")
        let! _ = AuctionTests.PlaceBid(runtime, auction, { Amount = 100m; Bidder = "alice" })

        runtime.Time.Advance(TimeSpan.FromSeconds(29.0))
        let lateBid = auction.PlaceBid({ Amount = 110m; Bidder = "bob" }, CancellationToken.None)
        do! runtime.RunToIdle()
        let! result = lateBid
        Assert.Equal(BidResult.Accepted, result)

        runtime.Time.Advance(TimeSpan.FromSeconds(2.0))
        do! runtime.RunToIdle()

        let! state = AuctionTests.ReadState(runtime, auction)
        Assert.Equal(AuctionState.Open, state)
        let! data = AuctionTests.ReadData(runtime, auction)
        Assert.Equal("bob", data.HighBidder)
    }

    [<Fact>]
    member this.State_machine_has_no_structural_defects() =
        ActorStateMachine.Analyze<AuctionActor>().AssertNoStructuralDefects()
        let defects = ActorStateMachine.Analyze<BadAuctionActor>().StructuralDefects
        Assert.Contains(defects, fun defect -> defect.Contains("unreachable", StringComparison.Ordinal))
        Assert.False(CoyoteBridge.IsEnabled)